pkg/pillar/cmd/zfsmanager/handlediskconfig.go

```diff
@@ -0,0 +1,136 @@
+	ctx := ctxArg.(*zfsContext)
+	select {
+	case ctx.disksProcessingTrigger <- struct{}{}:
+	default:
+		log.Functionf("handleDiskConfigRestart: disksProcessingTrigger already triggered")
+	}
+}
+
+func processDisksTask(ctx *zfsContext) {
+
+	disksProcessingTicker := time.NewTicker(disksProcessingInterval)
+
+	wdName := agentName + "diskstask"
+	stillRunning := time.NewTicker(stillRunningInterval)
+	ctx.ps.StillRunning(wdName, warningTime, errorTime)
+
+	for {
+		select {
+		case <-disksProcessingTicker.C:
+			if err := processDisks(ctx); err != nil {
+				log.Errorf("processDisks error: %s", err)
+			}
+		case <-ctx.disksProcessingTrigger:
+			if err := processDisks(ctx); err != nil {
+				log.Errorf("processDisks error: %s", err)
+			}
+		case <-stillRunning.C:
+		}
+		ctx.ps.StillRunning(wdName, warningTime, errorTime)
+	}
+}
+
+func processDisks(ctx *zfsContext) error {
+	if vault.ReadPersistType() != types.PersistZFS {
+		return nil
+	}
+	disksConfig := ctx.subDiskConfig.GetAll()
+	if len(disksConfig) == 0 {
+		return nil
+	}
+	persistPool, err := libzfs.PoolOpen(vault.DefaultZpool)
+	if err != nil {
+		return fmt.Errorf("cannot open pool: %s", err)
+	}
+	defer persistPool.Close()
+	vdevTree, err := persistPool.VDevTree()
+	if err != nil {
+		return fmt.Errorf("cannot get vdev tree: %s", err)
+	}
+	for _, cfg := range disksConfig {
+		diskCfg := cfg.(types.EdgeNodeDiskConfig)
+		diskName := filepath.Base(diskCfg.Disk.Name)
+		if diskCfg.OldDisk != nil {
+			oldDiskName := filepath.Base(diskCfg.OldDisk.Name)
+			oldDevName, oldDevStat := getVdevNameStatus(vdevTree, oldDiskName)
+			if oldDevName != "" && diskCfg.Disk.Name != "" {
+				// if we found device oldDevName, replace it
+				log.Functionf("replacing %s with %s, old stat %s", oldDevName, diskCfg.Disk.Name, oldDevStat.State.String())
+				if stdout, err := zfs.ReplaceVDev(log, vault.DefaultZpool, oldDevName, diskCfg.Disk.Name); err != nil {
+					log.Errorf("cannot replace %s with %s: %s %s", oldDevName, diskCfg.Disk.Name, stdout, err)
+					continue
+				}
+			}
+		}
+		devName, devStat := getVdevNameStatus(vdevTree, diskName)
+		// found in pool
+		if devName != "" {
+			log.Functionf("zpool config disk %s, op %d, stat %s", devName, diskCfg.Config, devStat.State.String())
+			switch diskCfg.Config {
+			case types.EdgeNodeDiskConfigTypeUnspecified:
+				continue
+			case types.EdgeNodeDiskConfigTypeZfsOnline:
+				if zfs.GetZfsDeviceStatusFromStr(devStat.State.String()) == info.StorageStatus_STORAGE_STATUS_ONLINE {
+					continue
+				}
+				if err := persistPool.Online(true, devName); err != nil {
+					log.Errorf("cannot bring %s online: %s", devName, err)
+				}
+			case types.EdgeNodeDiskConfigTypeZfsOffline:
+				if zfs.GetZfsDeviceStatusFromStr(devStat.State.String()) == info.StorageStatus_STORAGE_STATUS_OFFLINE {
+					continue
+				}
+				if err := persistPool.Offline(true, devName); err != nil {
+					log.Errorf("cannot bring %s offline: %s", devName, err)
+				}
+			default:
```

Comment:
If this for the case of setting the state to Unused? I think that should be an explicit case and not part of a default case.

Reply:
Good point, modified